### PR TITLE
Added TFLAudioTensor

### PR DESCRIPTION
**Pull request by priankakariatyml**
Added TFLAudioTensor and tests.

### CODE REVIEW

File: tensorflow_lite_support/ios/task/audio/core/audio_tensor/sources/TFLAudioTensor.h [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+NS_ASSUME_NONNULL_BEGIN
+
+/** A wrapper class to store input audio used in on-device machine learning. */
+NS_SWIFT_NAME(AudioTensor)
+@interface TFLAudioTensor : NSObject
+
+/** Audio format specifying the number of channels and sample rate supported. */
+@property(nonatomic, readonly) TFLAudioFormat *audioFormat;
+
+@property(nonatomic, readonly) TFLRingBuffer *ringBuffer;
```
Comment (khanhlvg) [inline tensorflow_lite_support/ios/task/audio/core/audio_tensor/sources/TFLAudioTensor.h:28]: Please add docstring

Comment (khanhlvg) [reply tensorflow_lite_support/ios/task/audio/core/audio_tensor/sources/TFLAudioTensor.h:28]: Besides, I think we shouldn't expose the TFLRingBuffer directly but only expose the TFLFloatBuffer of that ring buffer. It is to align with the Java API.

File: tensorflow_lite_support/ios/task/audio/core/audio_tensor/sources/TFLAudioTensor.m [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+                          @"(channels, sampleRate) and buffer size as TFLAudioTensor."];
+    return NO;
+  }
+
+  return [self loadBuffer:floatBuffer offset:0 size:floatBuffer.size error:error];
+}
+
+- (BOOL)loadAudioRecord:(TFLAudioRecord *)audioRecord withError:(NSError **)error {
+  if (![self.audioFormat isEqual:audioRecord.audioFormat]) {
+    NSLog(@"hellllllllllllllllo");
```
Comment (khanhlvg) [inline tensorflow_lite_support/ios/task/audio/core/audio_tensor/sources/TFLAudioTensor.m:62]: remove this

**Review by khanhlvg — CHANGES_REQUESTED** [review]
I left some comments. PTAL. Thanks.

File: tensorflow_lite_support/ios/task/audio/core/audio_tensor/sources/TFLAudioTensor.h (hunk 1 above)
Comment (priankakariatyml) [reply tensorflow_lite_support/ios/task/audio/core/audio_tensor/sources/TFLAudioTensor.h]: I have made the ring buffer a private instance variable and added a `TFLFloatBuffer` and bufferSize property to get the elements

File: tensorflow_lite_support/ios/task/audio/core/audio_tensor/sources/TFLAudioTensor.h [hunk 3]
```diff
@@ -25,7 +25,14 @@ NS_SWIFT_NAME(AudioTensor)
...
 @property(nonatomic, readonly) TFLAudioFormat *audioFormat;
 
-@property(nonatomic, readonly) TFLRingBuffer *ringBuffer;
+/**
+ * A copy of all the internal buffer elements in order with the most recent elements
+ * appearing at the end of `buffer.data`.
+ */
+@property(nonatomic, readonly) TFLFloatBuffer *buffer;
+
+/** Capacit of the `TFLAudioTensor` buffer in number of elements. */
```
Comment (khanhlvg) [inline tensorflow_lite_support/ios/task/audio/core/audio_tensor/sources/TFLAudioTensor.h:34]: typo: Capacity

File: tensorflow_lite_support/ios/task/audio/core/audio_tensor/sources/TFLAudioTensor.m [hunk 4]
```diff
@@ -16,13 +16,18 @@
...
 
 - (instancetype)initWithAudioFormat:(TFLAudioFormat *)format sampleCount:(NSUInteger)sampleCount {
   self = [self init];
   if (self) {
     _audioFormat = format;
-    _ringBuffer = [[TFLRingBuffer alloc] initWithBufferSize:sampleCount * format.channelCount];
+    
+    const NSInteger size = sampleCount * format.channelCount;
+    _ringBuffer = [[TFLRingBuffer alloc] initWithBufferSize:size];
+    _bufferSize = size;
```
Comment (khanhlvg) [inline tensorflow_lite_support/ios/task/audio/core/audio_tensor/sources/TFLAudioTensor.m:30]: nit: This is not used anywhere so I'd suggest removing this iVar and instead implement a getter for - (NSUInterger)bufferSize and return the `_ringBuffer` size directly.

File: tensorflow_lite_support/ios/task/audio/core/audio_tensor/sources/TFLAudioTensor.m [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+                               size:floatBuffer.size
+                              error:error];
+}
+
+- (BOOL)loadAudioRecordBuffer:(TFLFloatBuffer *)floatBuffer withError:(NSError **)error {
+  // Sample rate and channel count need not be checked here as they will be checked and reported
+  // when TFLAudioRecord which created this record buffer starts emitting wrong sized buffers.
+
+  // Checking buffer size makes sure that channel count and buffer size match.
+  if ([_ringBuffer size] != floatBuffer.size) {
```
Comment (khanhlvg) [inline tensorflow_lite_support/ios/task/audio/core/audio_tensor/sources/TFLAudioTensor.m:51]: Please use dot for property access.

**Comment by khanhlvg** [comment]
@priankakariatyml Mostly LGTM. I just left some minor comments. Thanks!

File: tensorflow_lite_support/ios/task/audio/core/audio_tensor/sources/TFLAudioTensor.h [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+ * of the `TFLAudioTensor`'s buffer size will be copied from the input buffer.
+ *
+ * @param sourceBuffer  A buffer of type `TFLFloatBuffer`. For multi-channel
+ * input, the array must be interleaved.
+ * @param offset Starting index in the source buffer from which elements should be copied.
+ * @param size The number of elements to be copied.
+ * @param error An optional error parameter populated with the reason for failure, if the internal
+ * buffer could not be loaded with the source buffer.
+ *
+ * @return A boolean indicating if the load operation succeded.
```
Comment (schmidt-sebastian) [inline tensorflow_lite_support/ios/task/audio/core/audio_tensor/sources/TFLAudioTensor.h:86]:
```suggestion
 * buffer could not be loaded with the source buffer.
 * @return A boolean indicating if the load operation succeded.
```

File: tensorflow_lite_support/ios/task/audio/core/audio_tensor/sources/TFLAudioTensor.h [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+/**
+ * Initializes a new `TFLAudioTensor` with a given `TFLAudioFormat` and sample count.
+ *
+ * @discussion The `TFLAudioTensor` stores data in a ring buffer of size sampleCount *
+ * TFLAudioFormat.channelCount.
+ *
+ * @param format An audio format of type `TFLAudioFormat`.
+ * @param sampleCount The number of samples `TFLAudioTensor` can store at any given
+ * time. The sampleCount provided will be used to calculate the buffer size of `TFLAudioTensor`
+ * bufferSize = format.channelCount * sampleCount.
```
Comment (schmidt-sebastian) [inline tensorflow_lite_support/ios/task/audio/core/audio_tensor/sources/TFLAudioTensor.h:46]:
```suggestion
 * (with `bufferSize = format.channelCount * sampleCount`).
 * @return A new instance of TFLAudioTensor with the given audio format and sample count.
```

File: tensorflow_lite_support/ios/task/audio/core/audio_tensor/sources/TFLAudioTensor.h [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+- (instancetype)initWithAudioFormat:(TFLAudioFormat *)format sampleCount:(NSUInteger)sampleCount;
+
+/**
+ * Convenience method to load the elements currently in the internal buffer of `TFLAudioRecord` into
+ * `TFLAudioTensor`.
+ *
+ * @discussion You must ensure that the  audio formats of `TFLAudioRecord` and the current
+ * `TFLAudioTensor` match.
+ * New data from the input buffer is appended to the end of the buffer by shifting out
+ * any old data from the beginning of the buffer if need be to make space. If the size of the new
```
Comment (schmidt-sebastian) [inline tensorflow_lite_support/ios/task/audio/core/audio_tensor/sources/TFLAudioTensor.h:59]:
```suggestion
 * any old data from the beginning of the buffer if needed to make space. If the size of the new
```

File: tensorflow_lite_support/ios/task/audio/core/audio_tensor/sources/TFLAudioTensor.h [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+ * @return A boolean indicating if the load operation succeded.
+ */
+- (BOOL)loadAudioRecord:(TFLAudioRecord *)audioRecord
+              withError:(NSError **)error NS_SWIFT_NAME(loadAudioRecord(audioRecord:));
+
+/**
+ * This function loads the internal buffer of `TFLAudioTensor` with the provided buffer.
+ *
+ * @discussion New data from the input buffer is appended to the end of the buffer by shifting out
+ * any old data from the beginning of the buffer if need be to make space. If the size of the new
```
Comment (schmidt-sebastian) [inline tensorflow_lite_support/ios/task/audio/core/audio_tensor/sources/TFLAudioTensor.h:75]:
```suggestion
 * any old data from the beginning of the buffer if needed to make space. If the size of the new
```

File: tensorflow_lite_support/ios/task/audio/core/audio_tensor/sources/TFLAudioTensor.h [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+ * data to be copied is more than the capacity of the buffer, only the most recent data
+ * of the `TFLAudioTensor`'s buffer size will be copied from the input buffer.
+ *
+ * @param sourceBuffer  A buffer of type `TFLFloatBuffer`. For multi-channel
+ * input, the array must be interleaved.
```
Comment (schmidt-sebastian) [inline tensorflow_lite_support/ios/task/audio/core/audio_tensor/sources/TFLAudioTensor.h:80]: Re-wrap comment please.

File: tensorflow_lite_support/ios/task/audio/core/audio_tensor/sources/TFLAudioTensor.h [hunk 11]
```diff
@@ -0,0 +1,95 @@
...
+ * @param offset Starting index in the source buffer from which elements should be copied.
+ * @param size The number of elements to be copied.
+ * @param error An optional error parameter populated with the reason for failure, if the internal
+ * buffer could not be loaded with the source buffer.
+ *
+ * @return A boolean indicating if the load operation succeded.
+ */
+- (BOOL)loadBuffer:(TFLFloatBuffer *)sourceBuffer
```
Comment (schmidt-sebastian) [inline tensorflow_lite_support/ios/task/audio/core/audio_tensor/sources/TFLAudioTensor.h:88]: Should this be `loadBufferFromSourceBuffer`? There is a discrepancy between 'Buffer` and 'sourceBuffer'.

**Review by khanhlvg — APPROVED** [review]

**Review by khanhlvg — APPROVED** [review]

File: tensorflow_lite_support/ios/task/audio/core/audio_tensor/sources/TFLAudioTensor.h (hunk 6 above)
Comment (priankakariatyml) [reply tensorflow_lite_support/ios/task/audio/core/audio_tensor/sources/TFLAudioTensor.h:87]: There was a mismatch between the spacing of the documentation of various functions in this file. My bad.
I intended to follow the format of [TFLInterpreter.h](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/objc/apis/TFLInterpreter.h). According to this file the space should be retained. I have updated the documentation of all functions in [TFLAudioTensor](https://github.com/tensorflow/tflite-support/pull/832/files/b57efe97d6bfe8710b52a02d2a3a7d6fdaaa2b35#diff-795194cf478a2a4f6881ace619b030aabdeb9f5d8c0b4b07bfd503a5c00bacab) to match the aforementioned file.
LMK if any other format needs to be followed.

**Review by schmidt-sebastian — APPROVED** [review]

**Pull request merged**